src/utils/crypto.rs

```diff
@@ -95,26 +96,84 @@ struct BundlerSimulationType {
 }
 
 /// ERC-4337 bundler userOperation schema
-#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, EthAbiCodec, EthAbiType)]
 #[serde(rename_all = "camelCase")]
 pub struct UserOperation {
-    pub sender: String,
-    pub nonce: String,
-    pub init_code: String,
-    pub call_data: String,
-    pub call_gas_limit: String,
-    pub verification_gas_limit: String,
-    pub pre_verification_gas: String,
-    pub max_fee_per_gas: String,
-    pub max_priority_fee_per_gas: String,
-    pub paymaster_and_data: String,
-    pub signature: String,
+    pub sender: EthersAddress,
+    pub nonce: U256,
+    pub init_code: Bytes,
+    pub call_data: Bytes,
+    pub call_gas_limit: U256,
+    pub verification_gas_limit: U256,
+    pub pre_verification_gas: U256,
+    pub max_fee_per_gas: U256,
+    pub max_priority_fee_per_gas: U256,
+    pub paymaster_and_data: Bytes,
+    pub signature: Bytes,
+}
+
+/// ERC-4337 bundler Packed userOperation schema for v07
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, EthAbiCodec, EthAbiType)]
+#[serde(rename_all = "camelCase")]
+pub struct PackedUserOperation {
+    pub sender: EthersAddress,
+    pub nonce: U256,
+    pub init_code: Bytes,
+    pub call_data: Bytes,
+    pub account_gas_limits: [u8; 32],
+    pub pre_verification_gas: U256,
+    pub gas_fees: [u8; 32],
+    pub paymaster_and_data: Bytes,
+    pub signature: Bytes,
+}
+
+impl UserOperation {
+    pub fn get_packed(&self) -> PackedUserOperation {
+        let account_gas_limits = concat_128(
+            self.verification_gas_limit.low_u128().to_be_bytes(),
+            self.call_gas_limit.low_u128().to_be_bytes(),
+        );
+
+        let gas_fees = concat_128(
+            self.max_priority_fee_per_gas.low_u128().to_be_bytes(),
+            self.max_fee_per_gas.low_u128().to_be_bytes(),
+        );
+
+        PackedUserOperation {
+            sender: self.sender,
+            nonce: self.nonce,
+            init_code: self.init_code.clone(),
+            call_data: self.call_data.clone(),
+            account_gas_limits,
+            pre_verification_gas: self.pre_verification_gas,
+            gas_fees,
+            paymaster_and_data: self.paymaster_and_data.clone(),
+            signature: self.signature.clone(),
+        }
+    }
+}
+
+fn concat_128(a: [u8; 16], b: [u8; 16]) -> [u8; 32] {
+    std::array::from_fn(|i| {
+        if let Some(i) = i.checked_sub(a.len()) {
+            b[i]
+        } else {
+            a[i]
+        }
+    })
```

Comment:
It took me a while to understand what this was doing; maybe something like this would be more readable:

```suggestion
    let mut result = [0; 32];
    result[..16].copy_from_slice(&a);
    result[16..].copy_from_slice(&b);
    result
```